### PR TITLE
[DATA-1649]/[DATA-1647] Add GetPointCloudMap collector and use streaming sync rpc for large files.

### PR DESCRIPTION
**Pull request by AaronCasas**
Use new streaming sync rpc for uploading binary capture files > 1MB.

# Testing
Confirmed that capture of point cloud maps works locally.
Confirmed that images are successfully uploaded to the backend using the streaming rpc.

### CODE REVIEW

File: services/datamanager/datasync/upload_data_capture_file.go [hunk 1]
```diff
@@ -20,22 +26,94 @@ func uploadDataCaptureFile(ctx context.Context, client v1.DataSyncServiceClient,
...
+		Type:             md.GetType(),
+		MethodParameters: md.GetMethodParameters(),
+		FileExtension:    md.GetFileExtension(),
+		Tags:             md.GetTags(),
 	}
-	_, err = client.DataCaptureUpload(ctx, ur)
-	if err != nil {
-		return err
+
+	// If it's a large binary file, we need to upload it in chunks.
```
Comment (agreenb) [inline services/datamanager/datasync/upload_data_capture_file.go:40]: Really great work making the streaming option readable and fit into the existing flow well by just checking the binary file size.

File: services/slam/collector.go [hunk 2]
```diff
@@ -39,6 +43,28 @@ func newGetPositionCollector(resource interface{}, params data.CollectorParams)
...
+		return nil, err
+	}
+
+	cFunc := data.CaptureFunc(func(ctx context.Context, _ map[string]*anypb.Any) (interface{}, error) {
+		f, err := slam.GetPointCloudMap(ctx)
+		if err != nil {
+			return nil, data.FailedToReadErr(params.ComponentName, getPointCloudMap.String(), err)
+		}
+
+		pcd, err := HelperConcatenateChunksToFull(f)
```
Comment (agreenb) [inline services/slam/collector.go:58]: Nice, glad SLAM already had a helper method to convert into bytes!

**Review by agreenb — COMMENTED** [review]
This looks fantastic, super excited for this addition! Just a few comments/nits and then looks ready to go. 🔥 

Thank you for adding the testing into the PR description — for the image testing to ensure that the streamed files were well-formed, did you reduce the chunk size and verify that it was using the streaming option?

File: services/datamanager/datacapture/data_capture_file.go [hunk 3]
```diff
@@ -240,7 +239,7 @@ func getFileTimestampName() string {
 // TODO DATA-246: Implement this in some more robust, programmatic way.
 func getDataType(methodName string) v1.DataType {
 	switch methodName {
-	case nextPointCloud, readImage:
+	case nextPointCloud, readImage, getPointCloudMap:
```
Comment (agreenb) [inline services/datamanager/datacapture/data_capture_file.go:242]: Whew good catch!

File: services/datamanager/builtin/sync_test.go [hunk 4]
```diff
@@ -332,31 +332,26 @@ func TestArbitraryFileUpload(t *testing.T) {
 		name                 string
 		manualSync           bool
 		scheduleSyncDisabled bool
-		serviceFail          bool
```
[the remark this answers is not in the document]
Comment (AaronCasas) [reply services/datamanager/builtin/sync_test.go]: Yup, added

File: services/datamanager/builtin/sync_test.go [hunk 5]
```diff
@@ -694,6 +813,15 @@ func (c mockDataSyncServiceClient) FileUpload(ctx context.Context, opts ...grpc.
 	return ret, nil
 }
 
+func (c mockDataSyncServiceClient) StreamingDataCaptureUpload(ctx context.Context, opts ...grpc.CallOption) (v1.DataSyncService_StreamingDataCaptureUploadClient, error) {
+	if c.fail.Load() {
```
[the remark this answers is not in the document]
Comment (AaronCasas) [reply services/datamanager/builtin/sync_test.go:829]: Good call, we should be. Added here and in FileUpload

File: services/datamanager/datasync/upload_data_capture_file.go [hunk 6]
```diff
@@ -20,22 +26,94 @@ func uploadDataCaptureFile(ctx context.Context, client v1.DataSyncServiceClient,
...
+	if md.GetType() == v1.DataType_DATA_TYPE_BINARY_SENSOR && f.Size() > MaxUnaryFileSize {
+		if len(sensorData) > 1 {
+			return errors.New("binary sensor data file with more than one sensor reading is not supported")
```
[the remark this answers is not in the document]
Comment (AaronCasas) [reply services/datamanager/datasync/upload_data_capture_file.go:43]: I think for now, I'd rather not change that logic as part of this PR. I don't have a strong opinion on doing it rdk side too

File: services/datamanager/datasync/upload_arbitrary_file.go [hunk 7]
```diff
@@ -11,7 +11,7 @@ import (
 )
 
 // UploadChunkSize defines the size of the data included in each message of a FileUpload stream.
-var UploadChunkSize = 64 * 1024
+var UploadChunkSize = 256 * 1024
```
[the remark this answers is not in the document]
Comment (AaronCasas) [reply services/datamanager/datasync/upload_arbitrary_file.go]: Reverted

File: services/datamanager/datasync/upload_arbitrary_file.go [hunk 8]
```diff
@@ -41,17 +41,14 @@ func uploadArbitraryFile(ctx context.Context, client v1.DataSyncServiceClient, f
...
+	if _, err := stream.CloseAndRecv(); err != nil {
 		return errors.Wrapf(err, "received error response while syncing %s", f.Name())
 	}
 
 	return nil
 }
 
 func sendFileUploadRequests(ctx context.Context, stream v1.DataSyncService_FileUploadClient, f *os.File) error {
-	//nolint:errcheck
-	defer stream.CloseSend()
```
Comment (agreenb) [inline services/datamanager/datasync/upload_arbitrary_file.go:54]: Why is this removed? Is this in order to synchronously have  `uploadArbitraryFile` and `uploadDataCaptureFile` in charge of closing the channel and getting the client response (via `CloseAndRecv`), rather than splitting that between the helper and main loop?

File: services/datamanager/builtin/sync_test.go [hunk 9]
```diff
@@ -425,29 +431,33 @@ func TestArbitraryFileUpload(t *testing.T) {
...
+					actData = append(actData, d.GetFileContents().GetData()...)
+				}
+				test.That(t, actData, test.ShouldResemble, fileContents)
+
+				// Validate file no longer exists.
+				waitUntilNoFiles(additionalPathsDir)
+				test.That(t, len(getAllFileInfos(additionalPathsDir)), test.ShouldEqual, 0)
+				test.That(t, dmsvc.Close(context.Background()), test.ShouldBeNil)
+			} else {
+				// Validate file still exists.
```
Comment (agreenb) [inline services/datamanager/builtin/sync_test.go:458]: For here and in the other tests, in this failure case, we should check that len(fileUploads) (or streaming uploads) == 0.

The file could technically still exist since we didn't call `waitUntilNoFiles`.

**Review by agreenb — COMMENTED** [review]
Just two more minor questions.

File: services/datamanager/datasync/upload_arbitrary_file.go (hunk 8 above)
Comment (AaronCasas) [reply services/datamanager/datasync/upload_arbitrary_file.go:54]: `CloseAndRecv` is the recommended/intended method for ending a client side stream ([docs](https://grpc.io/docs/languages/go/basics/#client-side-streaming-rpc-1). I believe this is because the stream ending and a response being received are supposed to always happen in this fixed order (since there is only one response message). In bidi streams where these can be interleaveed, [CloseSend](https://grpc.io/docs/languages/go/basics/#bidirectional-streaming-rpc-1) is used

**Review by agreenb — APPROVED** [review]
Fantastic work, really excited for SLAM to start using this! 🚀🚀🚀
